evm/vm/forks/sharding/windback_worker.py

```diff
@@ -0,0 +1,212 @@
+        self.collation_verifying_task[current_collation_hash] = task
+        self.chain_collations[head_collation_hash].append(current_collation_hash)
+
+    def get_chain_tasks(self, head_collation_hash):
+        # ignore tasks that are done
+        return [
+            self.collation_verifying_task[collation_hash]
+            for collation_hash in self.chain_collations[head_collation_hash]
+        ]
+
+    async def guess_head(self):
+        '''
+        Perform windback process.
+        Returns
+            the head collation hash, if it meets the windback condition. OR
+            None, if there's no qualified candidate head.
+        '''
+        start = time.time()
+
+        head_collation_hash = current_collation_hash = None
+        self.unfinished_verifying_tasks = []
+        candidate_heads = self.shard_tracker.fetch_candidate_heads_generator(
+            windback_length=self.vmc.config['WINDBACK_LENGTH']
+        )
+        while True:
+            try:
+                head_header = next(candidate_heads)
+            except StopIteration:
+                return None
+            head_collation_hash = head_header.hash
+            if not self.chain_validity[head_collation_hash]:
+                continue
+            current_collation_hash = head_collation_hash
+            for _ in range(self.vmc.config['WINDBACK_LENGTH'] + 1):
+                # if time is up
+                if self.is_time_up:
+                    return head_collation_hash
+                if current_collation_hash == GENESIS_COLLATION_HASH:
+                    break
+
+                # if a collation is an invalid head, set its descendants in the
+                # chain as invalid heads, and skip this chain
+                is_chain_of_current_collation_invalid = (
+                    (not self.chain_validity[current_collation_hash]) or
+                    (not self.collation_validity_cache.get(current_collation_hash, True))
+                )
+                if is_chain_of_current_collation_invalid:
+                    self.chain_validity[current_collation_hash] = False
+                    for collation_hash in self.chain_collations[head_collation_hash]:
+                        self.chain_validity[collation_hash] = False
+                    break
+                # process current collation  ##################################
+                # if the collation is checked before, then skip it
+                if current_collation_hash not in self.collation_validity_cache:
+                    descendants = copy.deepcopy(
+                        self.chain_collations[head_collation_hash]
+                    )
+                    coro = self.process_collation(
+                        head_collation_hash,
+                        current_collation_hash,
+                        descendants,
+                    )
+                    task = asyncio.ensure_future(coro)
+                    asyncio.wait(task, timeout=self.YIELDED_TIME)
```

Comment:
I'm still not very comfortable with asyncio, but I don't think `self.YIELDED_TIME` (and therefore `asyncio.wait`) is necessary here. `guess_head` should yield anytime a collation is actually downloaded. So I think `guess_head` should just wait for downloads to complete, not some arbitrary timeout.

Reply:
You're right `asyncio.wait(task, timeout=self.YIELDED_TIME)` is not necessary here.  I'm removing it.
Thank you